Source/inv.cpp

```diff
@@ -347,50 +489,88 @@ void CheckInvPaste(Player &player, Point cursorPosition)
+			if (player.InvGrid[testCell] != 0) {
+				int8_t iv = std::abs(player.InvGrid[testCell]);
+				if (it != 0) {
+					if (it != iv) {
+						// Found two different items that would be displaced by the held item, can't paste the item here.
+						return std::nullopt;
 					}
 				} else {
-					it = -iv;
+					it = iv;
 				}
 			}
-		} else {
-			// check that the item we're pasting only overlaps one other item (or is going into empty space)
-			unsigned originCell = static_cast<unsigned>(slot - SLOTXY_INV_FIRST);
-			for (unsigned rowOffset = 0; rowOffset < static_cast<unsigned>(itemSize.height * InventorySizeInSlots.width); rowOffset += InventorySizeInSlots.width) {
-				for (unsigned columnOffset = 0; columnOffset < static_cast<unsigned>(itemSize.width); columnOffset++) {
-					unsigned testCell = originCell + rowOffset + columnOffset;
-					// FindTargetSlotUnderItemCursor returns the top left slot of the inventory region that fits the item, we can be confident this calculation is not going to read out of range.
-					assert(testCell < sizeof(player.InvGrid));
-					if (player.InvGrid[testCell] != 0) {
-						int8_t iv = std::abs(player.InvGrid[testCell]);
-						if (it != 0) {
-							if (it != iv) {
-								// Found two different items that would be displaced by the held item, can't paste the item here.
-								return;
-							}
-						} else {
-							it = iv;
-						}
-					}
+		}
+	}
+
+	return it;
+}
+
+std::optional<int8_t> GetPrevItemId(int slot, const Player &player, const Size &itemSize)
+{
+	int8_t prevItemId = 0;
+	int cell_num = slot - SLOTXY_INV_FIRST; // Where in InvGrid
+	if (player.HoldItem._itype == ItemType::Gold) {
+		if (player.InvGrid[cell_num] != 0) {
+			int8_t item_cell_begin = player.InvGrid[cell_num];
+			if (item_cell_begin > 0) {
+				if (player.InvList[item_cell_begin - 1]._itype != ItemType::Gold) {
+					prevItemId = item_cell_begin;
 				}
+			} else {
+				prevItemId = -item_cell_begin;
 			}
 		}
-	} else if (il == ILOC_BELT) {
+	} else {
+		const std::optional<int8_t> overlappingItemId = CheckOverlappingItems(slot, player, itemSize, prevItemId);
+		if (!overlappingItemId) {
+			return std::nullopt;
+		}
+		prevItemId = *overlappingItemId;
+	}
+
+	return prevItemId;
+}
+
+void CheckInvPaste(Player &player, Point cursorPosition)
+{
+	const Size itemSize = GetInventorySize(player.HoldItem);
+
+	const int slot = FindTargetSlotUnderItemCursor(cursorPosition, itemSize);
+	if (slot == NUM_XY_SLOTS)
+		return;
+
+	auto const desired_loc = player.GetItemLocation(player.HoldItem);
+	const auto il = GetItemEquipType(slot, desired_loc, player);
+	if (il == ILOC_BELT) {
 		if (!CanBePlacedOnBelt(player, player.HoldItem))
 			return;
-	} else if (desiredIl != il) {
+	}
+
+	if ((il != ILOC_UNEQUIPABLE) && (desired_loc != il)) {
+		return;
+	}
+
+	auto const maybe_item_id = (il == ILOC_UNEQUIPABLE) ? GetPrevItemId(slot, player, itemSize) : 0;
+	if ((il == ILOC_UNEQUIPABLE) && !maybe_item_id) {
 		return;
 	}

```

Comment:
I'm not sure we are really winning anything by moving this up here. But maybe we could instead move it inside ChangeInvItem() since that is the only place this gets used. It would cause CalcPlrInv() to sometimes get called, but I don't think that is really an issue and if it is we can just have a return value for ChangeInvItem() that we check to see if we should exit the function early.
```suggestion
	std::optional<int8_t> maybe_item_id = 0;
	if (il == ILOC_UNEQUIPABLE) {
		maybe_item_id = GetPrevItemId(slot, player, itemSize);
		if (!maybe_item_id) return;
	}
```